zebra-state/src/service.rs

```diff
@@ -679,3 +690,43 @@ impl Service<Request> for StateService {
 pub fn init(config: Config, network: Network) -> BoxService<Request, Response, BoxError> {
     BoxService::new(StateService::new(config, network))
 }
+
+/// Check if zebra is following a legacy chain and return an error if so.
+pub fn legacy_chain_check<I>(
+    tip_height: block::Height,
+    ancestors: I,
+    network: Network,
+) -> Result<(), BoxError>
+where
+    I: Iterator<Item = Arc<Block>>,
+{
+    const MAX_BLOCKS_TO_CHECK: usize = 100;
+
+    // Only do the check if we have an Nu5 activation height.
+    if let Some(nu5_height) = NetworkUpgrade::Nu5.activation_height(network) {
+        // And only do the check if we are above Nu5 activation.
+        if tip_height >= nu5_height {
+            for (count, block) in ancestors.enumerate() {
+                if count >= MAX_BLOCKS_TO_CHECK {
+                    return Err("giving up after checking too many blocks".into());
+                }
+
+                // All transaction `network_upgrade` fields must be consistent with the block height.
+                block
+                    .check_transaction_network_upgrade_consistency(network)
+                    .map_err(|_| "inconsistent network upgrade found in transaction")?;
+
+                // If we find at least one transaction with a `network_upgrade` field we are ok.
```

Comment:
Explain design: How does this check detect a correct chain?
```suggestion
                // If we find at least one transaction with a valid `network_upgrade` field, the Zebra instance that
                // verified those blocks used the same network upgrade heights. (Up to this point in the chain.)
```

Reply:
https://github.com/ZcashFoundation/zebra/pull/2366/commits/8bb38d5ff2fa83dbabd5b1b7158148be0f475f1c